[Windows] Add video, HDR metadata, shader and swapchain infos to debug info OSD

thexai opened the pull request:
## Description
Add video, HDR metadata, shader and swapchain infos to debug info OSD.
Superseedes https://github.com/xbmc/xbmc/pull/19026

Unlike the previous PR, the main code has been moved outside of RenderManager and from there only two functions are called:

`m_pRenderer->GetDebugInfo`
`CServiceBroker::GetWinSystem()->GetDebugInfo`

To get the data from RendererBase and DeviceResouces. In RenderManager / DebugRenderer there are only minimal changes that allow adding more OSD text lines.

This has the advantage of being able to obtain much more detailed and platform exclusive data (e.g. flip discard / sequential, Windows HDR state) but cons: separate methods must be implemented for each platform (via override).

With this PR, the new infos is shown only for Windows platform. The rest will show the same 4 lines as before.

**The new infos are displayed separately with a new key map combination Alt+O (as a new feature). Ctrl+Shift+O shows same info as before even on Windows.**

I don't know if this PR can be included in v19 because there are many changes. Arguments to favor is that the changes are inside new `GetDebugInfo()` functions and the code never executes if the debug info OSD is not enabled. It is highly unlikely that any errors here could affect other parts of Kodi. Besides this, most of the information shown here is directly related to the new Windows HDR functionality added in v19 and therefore it would make sense to include it now.

## Motivation and Context
As comented in previus PR source video streams may have incomplete, incorrect, or missing metadata values and that can cause unexpected results if it's not know. Other infos as primaries or BT.2020 color space probably cannot be viewed anywhere else on Kodi. Dither and tone mapping values it's hard to check the status even looking at the debug log. Others swapchain parameters: flip discard, fullscreen exclusive can only be look in the debug log but it is more difficult to see.

Various people in the forums have expressed their concern to know if indeed when playing HDR content the output format is correct (high bit depth color) and if Windows HDR switches ON. e.g.: https://forum.kodi.tv/showthread.php?tid=359861

## How Has This Been Tested?
Runtime tested with different source streams and different's hardware/config's setups to check almost all different conditions.

Also various users are testing compilations that already includes it in the forum thread:
https://forum.kodi.tv/showthread.php?tid=359964

## Screenshots (if appropriate):
![screenshot00010](https://user-images.githubusercontent.com/58434170/111605295-7758ee00-87d6-11eb-9a98-12a81ee619d5.png)
This is a 4K HDR souce played in a PC monitor (1920x1200 no HDR). Can see that mapping es enabled and it's parameters (Hable, 1.0, 1004.70 nits) 1004.70 it's calculated from all metadata values: is the "average" value that is taken to apply tone map.
Since that display does not support film refresh rates is shown it's working at 59.94 Hz refresh but source is 23,976 fps


![screenshot00001](https://user-images.githubusercontent.com/58434170/111607130-41b50480-87d8-11eb-8c3e-3231a1603d13.png)
This is a 1080p SDR source played on a 4K TV. Can see output resolution is 3840 x 2160, this means it's up-scaled by Kodi. The output refresh rate 23.98 Hz matches source 23.976 fps shown in line one.

These are just two examples, obviously many different use cases.

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [X] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

thexai committed:
[Windows] Add video, HDR metadata, shader and swapchain infos to debug info OSD

enen92 commented:
@FernetMenta can you please take a look at this PR to see if it's right this time?

thexai commented:
Jenkins build this please

jjd-uk commented:
Maybe @lrusak could review the architecture of what is being done, as perhaps something similar could be done for the ongoing Linux HDR work.

lrusak commented:
I can give a bit of a review but I also think this should be deferred to V20

lrusak commented on xbmc/cores/VideoPlayer/VideoRenderers/DebugRenderer.h:
```diff
@@ -12,14 +12,21 @@
 
+#define MAX_LINES 10
+
+struct SInfo
```
What's the point of this structure?

lrusak commented on xbmc/cores/VideoPlayer/VideoRenderers/DebugRenderer.h:
```diff
@@ -12,14 +12,21 @@
 
 #include <string>
 
+#define MAX_LINES 10
```
define not namespaced at all pollutes the global landscape

lrusak commented on xbmc/cores/VideoPlayer/VideoRenderers/DebugRenderer.h:
```diff
@@ -32,7 +39,7 @@ class CDebugRenderer
 
-  std::string m_strDebug[4];
-  CDVDOverlayText *m_overlay[4];
+  std::string m_strDebug[MAX_LINES];
```
I think `std::array<std::string, MAX_LINES>` would be better or even better would be to use `std::vector<std::string>` as we don't have a defined size at compile time. This would also simplify some of the interface methods

lrusak commented on xbmc/cores/VideoPlayer/VideoRenderers/BaseRenderer.h:
```diff
@@ -92,6 +92,13 @@ class CBaseRenderer
   void SetVideoSettings(const CVideoSettings &settings);
 
+  // Get debug info from render buffer
+  virtual void GetDebugInfo(int idx,
```
I think a returning defined struct of `std::string` would be better suited here and more future proof.

lrusak commented on xbmc/windowing/WinSystem.h:
```diff
@@ -165,6 +165,9 @@ class CWinSystemBase
   virtual HDR_STATUS GetOSHDRStatus() { return HDR_STATUS::HDR_UNSUPPORTED; };
 
+  // Get debug info from video output stage
+  virtual void GetDebugInfo(std::string& info1, std::string& info2){};
```
`info1` and `info2` aren't great descriptors

lrusak commented on xbmc/windowing/win10/WinSystemWin10DX.h:
```diff
@@ -77,6 +77,9 @@ class CWinSystemWin10DX : public CWinSystemWin10, public CRenderSystemDX
   void SetHdrColorSpace(const DXGI_COLOR_SPACE_TYPE colorSpace) const;
 
+  // Get debug info from swapchain
+  void GetDebugInfo(std::string& info1, std::string& info2) override;
```
`info1` and `info2` aren't great descriptors

lrusak commented on xbmc/windowing/windows/WinSystemWin32DX.h:
```diff
@@ -79,6 +79,9 @@ class CWinSystemWin32DX : public CWinSystemWin32, public CRenderSystemDX
   void SetHdrColorSpace(const DXGI_COLOR_SPACE_TYPE colorSpace) const;
 
+  // Get debug info from swapchain
+  void GetDebugInfo(std::string& info1, std::string& info2) override;
```
`info1` and `info2` aren't great descriptors

lrusak commented on xbmc/cores/VideoPlayer/VideoRenderers/VideoShaders/WinVideoFilter.h:
```diff
@@ -70,6 +70,8 @@ class COutputShader : public CWinShader
                           bool hasLightMetadata, AVContentLightMetadata lightMetadata);
   void SetToneMapParam(int method, float param);
 
+  void GetDebugInfo(std::string& info);
```
why not just return a string here rather than passing by reference?

lrusak reviewed (changes requested):
I commend your effort to improve the debug interface, however, I think there are some areas that can be improved to remove some of the ambiguity that would be introduced in this PR.

In order to make it possible to be more platform agnostic I would recommend defining better structures of data that describe what they actually provide. This would allow the reader to more simply find what the methods are supposed to do and what they provide.

You'll have plenty of time to get this in for V20 and improve the interface along the way.

thexai commented:
Made most of the changes that @lrusak has requested.

Here I prefer to use `std::array` instead of `std::vector` because the number of lines will always be fixed in execution time and when it changes it's because someone is going to modify the code. It's also more efficient in memory occupation and access time. Let's not forget this is a debugging info... it's not worth wasting more CPU cycles just to make the code more stylish. 

Sometimes uglier code (old C style with fixed length data and structures) is more efficient... 😄

thexai commented:
I have made changes to use build-in ffmpeg methods to obtain pixel format, primaries and transfer readable names. 

This greatly simplifies the code.

lrusak commented:
I still think using a `struct` with a defined interface would be better than just an array of strings. This would improve readability for other devs.

My other issue is that the windowing part is very windows centric. Other platforms don't really have the notion of a `swapchain` and even though you can define whatever you want in the string it may be confusing for people wanting to implement it for other platforms.

The same concern is relevant to the shaders debug info.

I'm not sure if we should just allow platforms to define whatever debug strings they want or if we should be specific about what they should provide. I'd argue for the later as it would provide a more uniform debug overlay across platforms.

thexai commented:
It is precisely for this reason that I do not like a well-defined fixed structure. It is better to use more generic strings and that each platform uses it to display the information that is relevant for that platform.

e.g. In Linux there may not be a "swapchain" but an information line could be used to show the data of the HDR capabilities of the connected display (EDID).

If you notice before I had defined the "info1" "info2" strings but I changed it because you said it was not great descriptors:

```
  // Get debug info from video output stage
  virtual void GetDebugInfo(std::string& info1, std::string& info2){};
```

> info1 and info2 aren't great descriptors

I think that we cannot prefix the exact meaning of each line of information from the kernel. It should be flexible and that each platform shows the information that makes the most sense using windowing overrides methods.

lrusak commented:
> It is precisely for this reason that I do not like a well-defined fixed structure. It is better to use more generic strings and that each platform uses it to display the information that is relevant for that platform.
> 

A well defined structure can still have room for platform specifics

thexai commented:
If you prefer, you can define the structure you want and the base methods in windowing and I will adapt to it. 
That is, you make the kernel and I only implement the Windows overrides.

lrusak commented:
> If you prefer, you can define the structure you want and the base methods in windowing and I will adapt to it.
> That is, you make the kernel and I only implement the Windows overrides.

Looking at the code this took a bit to infer what is being grabbed
``` c++
struct debugInfo
{
  std::string audio;
  std::string video;
  std::string player;
  std::string videoFormatInfo;
  std::string videoMetaPrimaryInfo;
  std::string videoMetaLightInfo;
  std::string shaderInfo;
  std::string renderingInfo;
  std::string windowingInfo;
};
```

I would go further to make structs for the platform specifics.

``` c++
struct shaderInfo
{
  bool toneMappingEnabled{false};
  int toneMappingMethod;
  float toneMappingParam{0.0f};
  bool lutEnabled{false};
  int lutSize;
  bool ditheringEnabled{false};
  int ditheringDepth;
};
```
``` c++
struct renderingInfo
{
  int bufferCount;
  int flipMode;
};
```
``` c++
struct windowingInfo
{
  int windowMode;
  int width;
  int height;
  float refreshRate;
  int bitDepth;
  bool limitedRange;
  int minColorRange;
  int maxColorRange;
  bool scanModeProgressive;
  int eotf;
  int colorSpace;
  bool hdrEnabled{false};
};
```

This may also require defining some enum (I didn't bother to look if we already define some of these things).

I hope this helps creating a proper platform agnostic readable and intuitive interface. This may be more involved than what you signed up for but Kodi is by nature a multiplatform media center.

thexai commented:
I appreciate that you took the time to detail the data structures. In spite of everything, I still think that is not the best option. If I am not wrong, you want to pass the raw data in these structures and build the final readable strings at the end in `RenderManager.cpp`.

Since original data is already in similar structures (although at different places) first is need copy original data to new structures and move up though windowing. At end, similar readable strings are formatted (one for each OSD line).

With my approach, readable strings are formatted at origin and pass up through windowing. 

This has some advantages:

1. **Performance**: because avoids copy raw data step. (This needs to be done for each frame because some metadata may change). If we want more performance, some caching mechanism can be used... but this complicates things and leads to point 2. If debug OSD gets heavier, it will probably end up consuming 10% of CPU on less powerful systems. Currently on my NUC8i3 already uses 2-3% of CPU. (It seems too much to me).

2. **More reliability**: the original data being used is taken directly. E.g for swapchain flags is read directly from swapchain memory structure `DXGI_SWAP_CHAIN_DESC1` It’s the same data that the O.S. uses to render, so the chance of incorrect information is minimal. If this data is copied to auxiliary structure then OSD debug can indicate a "false" data because a bug in code or something. The purpose of the debug function is to know exactly what is actually on the screen. It would not be permissible for it to show 10-bit color depth by mistake because the information that is "cached" does not correspond to what is being used at that moment. This would make it lose its sense as a debugging tool.

3. **More flexible**: different platforms has different functionality implemented. E.g Windows has HLG to PQ shader. If readable string is constructed at origin is easy add info’s that are specific of this platform. Moreover, add info’s of future implemented features (probably not implemented at all platforms or not at same time). Other example: I prefer "Windows HDR is ON" because the string is constructed in windows specific code and can be personalized instead of "O.S. HDR is ON" or something similar that is confusing. Probably Linux or Android has not specific settings to enable/disable HDR at O.S. level and do not make sense show this at all. Only output EOTF is important. In Windows is possible play SDR stream with “Windows HDR ON” and TV detects HDR10 signal but EOTF used is SDR (traditional gamma, not PQ). Or playback HDR10 stream with "Windows HDR off" using SDR EOTF and tone mapping.

At the moment I am not modifying anything and I leave the PR waiting to obtain more support from others who think like me or otherwise this idea will probably be abandoned...

FernetMenta commented:
I agree with @thexai . In particular performance should not be underestimated. Some time ago I moved the debug info from a GUI widget to an overlay that is directly handled by RenderManager because of performance. Dynamic info like VSync offset become useless if showing the dialog consumes too much resources.
I would stay with this approach and in addition make all the information configurable. Instead of a single heavily overloaded screen it should be possible to show only the information important for the current debug session. You never look at HDR data and i.e. vsync offset at the same time.

thexai commented:
> I would stay with this approach and in addition make all the information configurable. Instead of a single heavily overloaded screen it should be possible to show only the information important for the current debug session. You never look at HDR data and i.e. vsync offset at the same time.

I like that idea: I changed with new shortcut key Alt+O for new infos and Ctrl+Shift+O show's same info as before

![screenshot00006](https://user-images.githubusercontent.com/58434170/109385623-80e8e780-78f5-11eb-82f1-8a1f34d2b990.png)

In this way a maximum of 6 text lines are displayed.

jjd-uk commented:
@lrusak now that FernetMenta commented, on this would you still like to see if any other devs have any comments on the correct approach for this? Maybe @afedchin and @fritsch if they have any time??

thexai commented:
Improved first text line:

Added source resolution (W x H).
Changed pixel format to real source pixel format: 3d311va_vld => yuv420p....  (since this line is intended to be the source information).
"Interlaced" string only is show when source is interlaced (there is not enough space to always show "progressive" and anyway 4K sources are always progressive).

![screenshot00003](https://user-images.githubusercontent.com/58434170/111067501-e11a8480-84c4-11eb-8f42-b482ab280383.png)

> I also think it would be good to preallocate these strings using the reserve string method to avoid reallocations (I'm not sure why this isn't done already because they talk about performance and wanting to pass references).

AFAIK `reserve` only can be useful on strings that change length drastically e.g. 2 characters to 1200 characters. Avoids continuously shrink and reallocate memory. If string only changes some character more or less (how is the case) I think a fixed memory length is maintained, probably in multiples of 2^n bytes 128, 256, 512, 1024 until next steep is needed or previous size is sufficient. I don't see that in this case this can influence performance.

fritsch commented:
For progressive or interlaced maybe i and p is enough at the proper location.

thexai commented:
![screenshot00009](https://user-images.githubusercontent.com/58434170/111069068-9cdeb280-84cb-11eb-9553-8714219f0448.png)

fritsch commented:
Looks nice! Thanks much.

thexai commented:
Jenkins build this please

thexai commented:
I tried implement Irusak's indications about interface: packing variables into simple structures. I hope that now is better implemented although no functional changes or performance impact.

@lrusak can you re-review?
Also want @afedchin for Windows related things

Windows tagged as new feature is only on Windows (for now), although it will be easy to expand to other platforms and I'm sure someone will be interested in doing it sooner or later.

thexai commented:
I think it's finished now. 😂

I have made some changes to improve the interface a bit more. There are no functional changes or changes in the Windows part  since afechin reviewed the code.

I would like the final go-ahead from Irusak.

thexai commented:
Jenkins build this please

lrusak commented on xbmc/cores/VideoPlayer/VideoRenderers/DebugRenderer.cpp:
```diff
@@ -33,50 +33,121 @@ CDebugRenderer::~CDebugRenderer()
     m_overlay[3] = new CDVDOverlayText();
     m_overlay[3]->AddElement(new CDVDOverlayText::CElementText(m_strDebug[3]));
   }
+  if ("" != m_strDebug[4])
```
is this just a placeholder?

lrusak commented on xbmc/cores/VideoPlayer/VideoRenderers/DebugRenderer.cpp:
```diff
@@ -33,50 +33,121 @@ CDebugRenderer::~CDebugRenderer()
+    m_overlay[4] = new CDVDOverlayText();
+    m_overlay[4]->AddElement(new CDVDOverlayText::CElementText(m_strDebug[4]));
+  }
+  if ("" != m_strDebug[5])
```
same thing here?

thexai replied:
You are right and this code can be removed:

```c++
  if ("" != m_strDebug[4])
  {
    m_strDebug[4] = "";
    if (m_overlay[4])
      m_overlay[4]->Release();
    m_overlay[4] = new CDVDOverlayText();
    m_overlay[4]->AddElement(new CDVDOverlayText::CElementText(m_strDebug[4]));
  }
  if ("" != m_strDebug[5])
  {
    m_strDebug[5] = "";
    if (m_overlay[5])
      m_overlay[5]->Release();
    m_overlay[5] = new CDVDOverlayText();
    m_overlay[5]->AddElement(new CDVDOverlayText::CElementText(m_strDebug[5]));
  } 
```

And change `for` to ` i < 4` as debug info player only has 4 lines (is not need render empty lines).

lrusak commented on xbmc/windowing/win10/WinSystemWin10DX.cpp:
```diff
@@ -196,3 +196,8 @@ void CWinSystemWin10DX::SetHdrColorSpace(const DXGI_COLOR_SPACE_TYPE colorSpace)
+DEBUG_INFO_RENDER CWinSystemWin10DX::GetDebugInfo()
+{
+  return m_deviceResources->GetDebugInfo();
+}
```
no new line at end of file

thexai replied:
Done!

lrusak commented on xbmc/windowing/windows/WinSystemWin32DX.cpp:
```diff
@@ -421,3 +421,8 @@ void CWinSystemWin32DX::SetHdrColorSpace(const DXGI_COLOR_SPACE_TYPE colorSpace)
+DEBUG_INFO_RENDER CWinSystemWin32DX::GetDebugInfo()
+{
+  return m_deviceResources->GetDebugInfo();
+}
```
same thing here

lrusak reviewed (commented):
I just commented on some minors. Overall it looks a lot better and is much more readable now. Thanks for taking the time to make the changes.

lrusak reviewed (approved):
LGTM. Thanks for taking the time to make the requested changes.

thexai commented:
Now that it is approved I want to propose backport it to Matrix. I think the only reason it was not included is because it could not be finished in time with sufficient quality.

Now that it is ready it should not be a problem to include it in v19.1 as it is a debugging information closely related to HDR10 and the new tone mapping methods that was introduced in v19.0 (Windows).

lrusak commented:
> Now that it is approved I want to propose backport it to Matrix. I think the only reason it was not included is because it could not be finished in time with sufficient quality.
> 
> Now that it is ready it should not be a problem to include it in v19.1 as it is a debugging information closely related to HDR10 and the new tone mapping methods that was introduced in v19.0 (Windows).

that's up to @DaveTBlake

KarellenX commented:
If any errors, let me know

https://kodi.wiki/view/Player_process_info#HDR_Debug_Info

thexai commented:
Maybe better name is "Video Debug Info" as is not HDR specific but all video source and render infos.

And here https://kodi.wiki/view/Keyboard_controls is named "HDR process info"  ---->  Video Debug Info

Finally here https://kodi.wiki/view/Video_playback can be added a reference similar to  "7 Player Debug Info"

Thank you to document this!

KarellenX commented:
Thanks @thexai Done x3

mitzsch commented:
Some years later and Linux HDR support got better...

Can this be adapted to Linux? Or at least show the source metadata values of the file in this debug overlay on other platforms?
